### ISSUE TITLE
Kaspersky product(?) causes crash inside project view counter

### ISSUE THREAD

**Issue by apple502j** [LLK/scratch-www#4326]
There are some reports that green flag "doesn't work" and projects don't run. Luckily, one of the reporters ([xcalibur98](https://scratch.mit.edu/discuss/post/4334999/)) gave me a traceback:

```
Uncaught TypeError: Cannot read property ‘slice’ of null
    at XMLHttpRequest.RequestSend (main.js:2631)
    at XMLHttpRequest.xhrSendProcessor (main.js:2653)
    at XMLHttpRequest.window.XMLHttpRequest.send (main.js:2562)
    at XMLHttpRequest.send (projects.bundle.js:31)
    at i (common.bundle.js:16)
    at a (common.bundle.js:16)
    at r (common.bundle.js:6)
    at common.bundle.js:6
    at common.bundle.js:6
    at Object.get (common.bundle.js:6)
    at Object.use (common.bundle.js:6)
    at e.exports (common.bundle.js:6)
    at projects.bundle.js:47
    at common.bundle.js:38
    at Object.logProjectView (projects.bundle.js:171)
    at t.value (projects.bundle.js:171)
    at t.n (common.bundle.js:6)
    at t.i.emit (projects.bundle.js:68)
    at t.<anonymous> (projects.bundle.js:146)
    at t.i.emit (projects.bundle.js:68)
    at t.value (projects.bundle.js:119)
    at t.value (projects.bundle.js:146)
    at t.value (projects.bundle.js:166)
    at n (projects.bundle.js:52)
    at Object.u (common.bundle.js:22)
    at Object.invokeGuardedCallback (common.bundle.js:30)
    at Object.invokeGuardedCallbackAndCatchFirstError (common.bundle.js:30)
    at m (common.bundle.js:22)
    at y (common.bundle.js:22)
    at b (common.bundle.js:22)
    at Array.forEach (<anonymous>)
    at g (common.bundle.js:22)
    at $ (common.bundle.js:22)
    at Gn (common.bundle.js:30)
    at Ce (common.bundle.js:22)
    at batchedUpdates (common.bundle.js:22)
    at Q (common.bundle.js:22)
    at Le (common.bundle.js:22)
```

Because the file is minified, the traceback is not very useful - however, it told me that "logProjectView" - https://github.com/LLK/scratch-www/blob/develop/src/redux/preview.js#L1005 - is crashing.

Because the function uses callback instead of then-chain, we are not sure which XHR is crashing - it is either https://github.com/LLK/scratch-www/blob/develop/src/lib/jar.js#L70 or https://github.com/LLK/scratch-www/blob/develop/src/lib/api.js#L65 - I think it's former one, though. I asked the reporter to tell me if there is an request to views API.

Also, when I google "xhrSendProcessor" there is no result, which is concerning. Also, "main.js" is not a file on the Scratch server. However, because multiple people are reporting similar bugs, I am sure this is not just their computer.

## Repros
- https://scratch.mit.edu/discuss/post/4322471/ "Projects never starting when Green Flag pressed", My browser / operating system: Windows NT 10.0, Chrome 84.0.4147.125, No Flash version detected, the reporter says it happened on Edge and Firefox
- https://scratch.mit.edu/discuss/post/4267585/ "Green Flag Won't Work AT ALL", My browser / operating system: Windows NT 10.0, Chrome 84.0.4147.89, No Flash version detected, the reporter says they're actually using Chromium Edge, Ctrl-F5 didn't work, Green flag overlay (on scratch-gui stage) disappeared correctly
- https://scratch.mit.edu/discuss/post/4311609/ "Problem with green flag not working" - "I tried refreshing it, logging out, and logging back in, deleting my history." My browser / operating system: Windows NT 10.0, Chrome 84.0.4147.105, No Flash version detected
- https://scratch.mit.edu/discuss/post/4298161/ "The green flag won't work."
- https://scratch.mit.edu/discuss/post/4228909/ "Green Flag Broken??????"

## Debugging - August 19
Thanks to a video I identified that this is a bug on Kaspersky product not following XMLHttpRequest specs, which says the first argument to `XMLHttpRequest.send` can be `null`. Kaspersky code checks `arguments.length > 0` and if it's true it calls `arguments[0].slice`, causing error (because `arguments[0]` is null)

**Comment by R4356th** [comment]
Can't repro on Chrome latest version on Android 7 or Windows 10.

**Comment by GrahamSH-LLK** [comment]
Can't repro with 
Google Chrome: Version 84.0.4147.110 (Official Build)  (64-bit)
Platform: 13099.85.0 (Official Build) stable-channel atlas
Channel: stable-channel
Firmware Version: Google_Atlas.11827.46.0

**Comment by apple502j** [comment]
According to [the video by the repro-er](https://www.youtube.com/watch?v=OoDkwMagQb4), the problematic "main.js" mentions Kaspersky's anti-virus software. Maybe it has something to do with it.

**Comment by ljy-002** [comment]
What？

**Comment by ljy-002** [comment]
> Can't repro on Chrome latest version on Android 7 or Windows 10.

What，I can

**Comment by apple502j** [comment]
@ljy-002 screenshots?

**Comment by ljy-002** [comment]
> @ ljy-002截图？

![image](https://user-images.githubusercontent.com/63292034/90616072-8925e300-e23f-11ea-9bea-298da6d24d3c.png)
This is the first URL he said

**Comment by ljy-002** [comment]
> @ ljy-002截图？

but YouTube……：
![image](https://user-images.githubusercontent.com/63292034/90616188-b5d9fa80-e23f-11ea-8585-290e3273dca0.png)

**Comment by R4356th** [comment]
Do you live in China? YouTube is banned there.

**Comment by R4356th** [comment]
@ljy-002, @apple502j asked for a screenshot of the bug not the forum post.

**Comment by apple502j** [comment]
Debugging:
1) Kaspersky - they assume `arguments[0]` can have properties as long as `arguments.length > 0` - which is incorrect because `arguments[0]` can be null.
2) xhr (npm library) - the argument to `send()` is the variable `body`, defined at https://github.com/naugtur/xhr/blob/0dc970078e532f6a74e059a92b2b2d9744e8b324/index.js#L151 as `options.body || options.data || null`
3) Scratch - when XHRing, neither `body` nor `data` is specified.

Short answer: *can you stop using XHR and use fetch()?*
Long answer: maybe default to empty string if it works?

**Comment by apple502j** [comment]
This issue is fixed - install Kaspersky Safe Kids version 1.0.5.6390
